alacritty/src/renderer/text/gles2.rs

```diff
@@ -0,0 +1,478 @@
+
+            // Elements buffer.
+            gl::BindBuffer(gl::ELEMENT_ARRAY_BUFFER, ebo);
+            gl::BufferData(
+                gl::ELEMENT_ARRAY_BUFFER,
+                (vertex_indices.capacity() * size_of::<u16>()) as isize,
+                vertex_indices.as_ptr() as *const _,
+                gl::STATIC_DRAW,
+            );
+
+            // Vertex buffer.
+            gl::BindBuffer(gl::ARRAY_BUFFER, vbo);
+            gl::BufferData(
+                gl::ARRAY_BUFFER,
+                (BATCH_MAX * size_of::<TextVertex>()) as isize,
+                ptr::null(),
+                gl::STREAM_DRAW,
+            );
+
+            let mut index = 0;
+            let mut size = 0;
+
+            macro_rules! add_attr {
+                ($count:expr, $gl_type:expr, $type:ty) => {
+                    gl::VertexAttribPointer(
+                        index,
+                        $count,
+                        $gl_type,
+                        gl::FALSE,
+                        size_of::<TextVertex>() as i32,
+                        size as *const _,
+                    );
+                    gl::EnableVertexAttribArray(index);
+
+                    #[allow(unused_assignments)]
+                    {
+                        size += $count * size_of::<$type>();
+                        index += 1;
+                    }
+                };
+            }
+
+            // Cell coords.
+            add_attr!(2, gl::SHORT, i16);
+
+            // Glyph coords.
+            add_attr!(2, gl::SHORT, i16);
+
+            // UV.
+            add_attr!(2, gl::FLOAT, u32);
+
+            // Color and bitmap color.
+            //
+            // These are packed together because of an OpenGL driver issue on macOS, which caused a
+            // `vec3(u8)` text color and a `u8` for glyph color to cause performance regressions.
+            add_attr!(4, gl::UNSIGNED_BYTE, u8);
+
+            // Background color.
+            add_attr!(4, gl::UNSIGNED_BYTE, u8);
+
+            // Cleanup.
+            gl::BindVertexArray(0);
+            gl::BindBuffer(gl::ELEMENT_ARRAY_BUFFER, 0);
+            gl::BindBuffer(gl::ARRAY_BUFFER, 0);
+        }
+
+        Ok(Self {
+            program,
+            vao,
+            vbo,
+            ebo,
+            atlas: vec![Atlas::new(ATLAS_SIZE)],
+            batch: Batch::new(),
+            current_atlas: 0,
+            active_tex: 0,
+        })
+    }
+}
+
+impl Drop for Gles2Renderer {
+    fn drop(&mut self) {
+        unsafe {
+            gl::DeleteBuffers(1, &self.vbo);
+            gl::DeleteBuffers(1, &self.ebo);
+            gl::DeleteVertexArrays(1, &self.vao);
+        }
+    }
+}
+
+impl<'a> TextRenderer<'a> for Gles2Renderer {
+    type RenderApi = RenderApi<'a>;
+    type RenderBatch = Batch;
+    type Shader = TextShaderProgram;
+
+    fn program(&self) -> &Self::Shader {
+        &self.program
+    }
+
+    fn with_api<'b: 'a, F, T>(&'b mut self, _: &'b SizeInfo, func: F) -> T
```

Comment:
Bit unfortunate this method still seems to basically be completely duplicated? Though I don't think there's any great way to solve this.

Reply:
Yeah, I don't see how it could be solved, given the types are just different.